src/Microsoft.IdentityModel.Protocols.SignedHttpRequest/SignedHttpRequestHandler.cs

```diff
@@ -0,0 +1,1417 @@
+        {
+            if (payload == null)
+                throw LogHelper.LogArgumentNullException(nameof(payload));
+
+            var signedHttpRequestCreationTime = DateTime.UtcNow.Add(signedHttpRequestDescriptor.SignedHttpRequestCreationParameters.TimeAdjustment);
+            payload.Add(SignedHttpRequestClaimTypes.Ts, (long)(signedHttpRequestCreationTime - EpochTime.UnixEpoch).TotalSeconds);
+        }
+
+        /// <summary>
+        /// Adds the 'm' claim to the <paramref name="payload"/>.
+        /// </summary>
+        /// <param name="payload">HttpRequest payload represented as a <see cref="Dictionary{TKey, TValue}"/>.</param>
+        /// <param name="signedHttpRequestDescriptor">A structure that wraps parameters needed for SignedHttpRequest creation.</param>
+        /// <remarks>
+        /// This method will be executed only if <see cref="SignedHttpRequestCreationParameters.CreateM"/> is set to <c>true</c>.
+        /// </remarks>   
+        protected virtual void AddMClaim(Dictionary<string, object> payload, SignedHttpRequestDescriptor signedHttpRequestDescriptor)
+        {
+            if (payload == null)
+                throw LogHelper.LogArgumentNullException(nameof(payload));
+
+            var httpMethod = signedHttpRequestDescriptor.HttpRequestData.Method;
+
+            if (string.IsNullOrEmpty(httpMethod))
+                throw LogHelper.LogArgumentNullException(nameof(signedHttpRequestDescriptor.HttpRequestData.Method));
+
+            if (!httpMethod.ToUpper().Equals(httpMethod, StringComparison.Ordinal))
+                throw LogHelper.LogExceptionMessage(new SignedHttpRequestCreationException(LogHelper.FormatInvariant(LogMessages.IDX23002, httpMethod)));
+
+            payload.Add(SignedHttpRequestClaimTypes.M, httpMethod);
+        }
+
+        /// <summary>
+        /// Adds the 'u' claim to the <paramref name="payload"/>.
+        /// </summary>
+        /// <param name="payload">HttpRequest payload represented as a <see cref="Dictionary{TKey, TValue}"/>.</param>
+        /// <param name="signedHttpRequestDescriptor">A structure that wraps parameters needed for SignedHttpRequest creation.</param>
+        /// <remarks>
+        /// This method will be executed only if <see cref="SignedHttpRequestCreationParameters.CreateU"/> is set to <c>true</c>.
+        /// </remarks>  
+        protected virtual void AddUClaim(Dictionary<string, object> payload, SignedHttpRequestDescriptor signedHttpRequestDescriptor)
+        {
+            if (payload == null)
+                throw LogHelper.LogArgumentNullException(nameof(payload));
+
+            var httpRequestUri = signedHttpRequestDescriptor.HttpRequestData.Uri;
+
+            if (httpRequestUri == null)
+                throw LogHelper.LogArgumentNullException(nameof(signedHttpRequestDescriptor.HttpRequestData.Uri));
+
+            if (!httpRequestUri.IsAbsoluteUri)
```

Comment:
Should the utility EnsureAbsoluteUri be used here as well?

Reply:
The 'u' claim represents an http host component, which requires presence of an absolute URI.